litr-filters/src/main/java/com/linkedin/android/litr/filter/audio/AudioOverlayFilter.kt

```diff
@@ -0,0 +1,211 @@
+        var inputTag = MediaCodec.INFO_TRY_AGAIN_LATER
+        while (inputTag < 0) {
+            inputTag = decoder.dequeueInputFrame(-1)
+        }
+        val decoderInputFrame = decoder.getInputFrame(inputTag)
+            ?: throw TrackTranscoderException(TrackTranscoderException.Error.NO_FRAME_AVAILABLE)
+
+        val bytesRead = mediaSource.readSampleData(decoderInputFrame.buffer!!, 0)
+        val sampleTime = mediaSource.sampleTime
+        val sampleFlags = mediaSource.sampleFlags
+
+        if (bytesRead == 0) {
+            // nothing was read from the source
+            return null
+        }
+
+        // send the frame to decoder and advance the source position
+        decoderInputFrame.bufferInfo[0, bytesRead, sampleTime] = sampleFlags
+        decoder.queueInputFrame(decoderInputFrame)
+        mediaSource.advance()
+
+        // get the decoder output (decoded frame)
+        var outputTag = MediaCodec.INFO_TRY_AGAIN_LATER
+        while (outputTag < 0) {
+            outputTag = decoder.dequeueOutputFrame(-1)
+        }
+
+        val decoderOutputFrame = decoder.getOutputFrame(outputTag)
+            ?: throw TrackTranscoderException(TrackTranscoderException.Error.NO_FRAME_AVAILABLE)
+
+        // resize decoded overlay frame to match the target frame it will be applied to
+        val sourceSampleCount = decoderOutputFrame.bufferInfo.size / (BYTES_PER_SAMPLE * overlayChannelCount)
+        val estimatedTargetSampleCount = ceil(sourceSampleCount * samplingRatio).toInt()
+        val targetBufferCapacity = estimatedTargetSampleCount * channelCount * BYTES_PER_SAMPLE
+
+        val processedBuffer = bufferPool.get(targetBufferCapacity)
+        val processedFrame = Frame(outputTag, processedBuffer, MediaCodec.BufferInfo())
+        audioProcessor?.processFrame(decoderOutputFrame, processedFrame)
+
+        decoder.releaseOutputFrame(outputTag, false)
+
+        return processedFrame
+    }
+
+    private fun renderOverlay(frameBuffer: ByteBuffer) {
+        while (frameBuffer.remaining() > 0) {
+            renderQueue.peek()?.let { overlayBuffer ->
+                applyOverlaySample(
+                    frameBuffer,
+                    overlayBuffer,
+                    min(frameBuffer.remaining(), overlayBuffer.remaining())
+                )
+
+                if (overlayBuffer.remaining() == 0) {
+                    renderQueue.removeFirst()
+                    bufferPool.put(overlayBuffer)
+                }
+            }
+        }
+
+        // reset the position back to 0 by flipping the frame to "read" mode
+        frameBuffer.flip()
+    }
+
+    private fun applyOverlaySample(frameBuffer: ByteBuffer, overlayBuffer: ByteBuffer, byteCount: Int) {
+        repeat(byteCount / 2) {
+            val mixedValue = (frameBuffer.short + overlayBuffer.short) / 2
+            frameBuffer.putShort(frameBuffer.position() - 2, mixedValue.toShort())
+        }
+    }
+}
```

Comment:
It works, but it would need to be improved, e.g., if one track has normal volume, another is quiet at that time, the simple average algorithm will quieter the original one which user can tell easily and sometimes feel weird.

Reply:
Improved. Not manipulating volume anymore, simply overlaying two tracks together.